ipykernel/kernelapp.py

```diff
@@ -367,7 +367,7 @@ def close(self):
             if socket and not socket.closed:
                 socket.close()
         self.log.debug("Terminating zmq context")
-        # FIXME
+        # FIXME: this line breaks shutdown
         # self.context.term()
```

Comment:
Not sure why terminating the ZMQ context creates issues here.

Reply:
Maybe related: https://github.com/jupyter-server/jupyter_server/issues/677#issuecomment-1033597413

Reply:
Thanks @blink1073, I can see that currently, [we don't close the heartbeat socket before terminating the ZMQ context](https://github.com/ipython/ipykernel/blob/d6744f9e423dacc6b317b1d31805304e89cbec5d/ipykernel/heartbeat.py#L112), but the other way around: we wait for the context to be terminated before closing the socket.
Maybe this can be the root cause, @minrk ?

Reply:
Context.term blocks until all sockets are closed and all unsent messages are delivered or discarded. A socket may close after the context if it is done from another thread. But it can't be done from a coroutine, for instance.

The things to check when a context hangs are always:

1. are there sockets that haven't been closed, and
2. was LINGER set on those sockets so they discard messages eventually

You can debug leftover sockets with `self.context._sockets` which is a set of weakrefs to sockets (this is private and subject to change, but works for debugging).

The heartbeat socket [doesn't share this context](https://github.com/ipython/ipykernel/blob/d6744f9e423dacc6b317b1d31805304e89cbec5d/ipykernel/kernelapp.py#L337), so I don't think that's relevant.